### ISSUE TITLE
multi-output layer(different labels) training in NNET3

### ISSUE THREAD

**Issue by cathryn0131** [kaldi-asr/kaldi#2300]
@danpovey 
hi,again
I am training an NNET(something little bit like adversarial training). After checking the NNET3 code, I still havn't found any solutions, so could you please help check and show a way, thanks in advance.
     The NNET has two output layers(label is utterance and spkid separately), the cost function equals the sum of the cost of two output layers.
      I checked the chain model, but the output and output-xent should have same dimension and labels. I also checked multi-lingual, but it modified the output name to different language name and calculate with posterior and readjust priors, which seems not working too. So is there any suggestion on how to carry on the training? thanks very much
The NNET structure as planned initially is as following:
  input dim=41 name=input
  fixed-affine-layer name=lda input=Append(-3,-2,-1,0,1,2,3) affine-transform-file=exp/nnet3/tdnn_d/configs/lda
  relu-renorm-layer name=tdnn1 dim=1024
  relu-renorm-layer name=tdnn2 input=Append(-1,2) dim=1024 learning-rate-factor=0.1
  relu-renorm-layer name=tdnn3 input=Append(-3,3) dim=1024
  relu-renorm-layer name=tdnn4 input=Append(-3,3) dim=1024
  relu-renorm-layer name=tdnn5 input=Append(-7,2) dim=1024
  relu-renorm-layer name=tdnn6 dim=1024
  relu-batchnorm-layer name=tdnn7 input=Append(-2,-1,0,1,2) dim=1952
  relu-batchnorm-layer name=tdnn8 input=Append(-2,0,2) dim=512
  relu-batchnorm-layer name=tdnn9 input=Append(-3,0,3) dim=512
  stats-layer name=stats config=mean+stddev(0:1:1:10000)
  relu-batchnorm-layer name=tdnn10 dim=512 input=stats learning-rate-factor=0.1
  relu-batchnorm-layer name=tdnn11 dim=512
  output-layer name=output-spk include-log-softmax=true dim=1900
  output-layer name=output-utt input=tdnn6 dim=1952 max-change=1.5 presoftmax-scale-file=exp/nnet3/tdnn_d/configs/p

p.s: input is utterance audio wav
      output-utt is the text for asr result, label for this layer is real text
      output-spk is the spkid for wav, label for this layer is spkid(tdnn7 to output-spk is xvector nnet)
      The labels for the two output layer are totally different

**Comment by danpovey** [comment]
This should really be on kaldi-help but I'll respond here this time.

You say it's adversarial but I don't see any mention of negating the derivatives from the speaker part.

What you want to do should be possible without coding, but there are some subtle issues involved and maybe some scripting work.

Before I put more effort into helping you I want to understand your motivation here.  Do you want to solve speaker-id or ASR?

**Comment by cathryn0131** [comment]
@dan
sorry for put it in wrong part.
I train this NNET to reduce the affect of  user voice characteristics.
the derivative of tdnn6 layer is the sum of deriv1 from output-utt(output-utt -> tdnn6) and deriv2 from tdnn7(output-spk -> tdnn11 -> tdnn10 ->tdnn9 -> tdnn8 -> tdnn7 -> tdnn6).
derivative(tdnn6) = deriv1 + coeff * deriv2
I will set the learn rate of some layers to 0 to keep is unchanged.
step 1 : set learning rate of tdnn11, tdnn10,tdnn9,tdnn8,tdnn7,tdnn6 to zero
step 2 : set learning rate of lda,tdnn1, tdnn2,tdnn3,tdnn4,tdnn5 to zero
and takes step 1 and step 2 several times

**Comment by danpovey** [comment]
So I assume 'coeff' would be negative--  Otherwise I don't see how it would reduce the effect of speaker characteristics and I don't see what is adversarial about it.
I guess this is similar to MTL as in https://arxiv.org/pdf/1703.02136.pdf.
However the improvements they got from it were extremely tiny and I don't believe it will be helpful at all in the context of a LF-MMI objective function.

Sorry, I think my position on this is going to be that it's not a good use of my time to work with you to implement this, because you will need too much help, and I don't think the idea is promising.
It would require coding a layer that would reverse the derivative (not too hard), and you'd probably want to have the speaker labels frame by frame, that would probably work better.  It could mostly, otherwise, be done without binary changes, except possibly to create egs with extra outputs.  (You'd combine both targets in the same NnetExample object). 

But I just don't think what you are describing will help.
